File: primes.js
'use strict';

// Sieve of Eratosthenes
function eratosthenes(limit) {
    let prms = [];
    if (limit >= 2) prms = [2];
    if (limit >= 3) {
        let sqrtlmt = (Math.sqrt(limit) - 3) >> 1;
        let lmt = (limit - 3) >> 1;
        let bfsz = (lmt >> 5) + 1;
        let buf = [];
        for (let i = 0; i < bfsz; i++)
            buf.push(0);
        for (let i = 0; i <= sqrtlmt; i++)
            if ((buf[i >> 5] & (1 << (i & 31))) == 0) {
                let p = i + i + 3;
                for (let j = (p * p - 3) >> 1; j <= lmt; j += p)
                    buf[j >> 5] |= 1 << (j & 31);
            }
        for (let i = 0; i <= lmt; i++)
            if ((buf[i >> 5] & (1 << (i & 31))) == 0)
                prms.push(i + i + 3);
    }
    return prms;
}

function isPrime(num) {
    if (num <= 1) return false;
    if (num <= 3) return true;
    if (((num % 2) == 0) || ((num % 3) == 0)) return false;
    for (let c = 5; c*c <= num; c += 6) 
        if (((num % c) == 0) || ((num % (c+2)) == 0)) return false;
    return true;
}

// Primos
let noPrimes = 0;
const min = 0, limit = 20000000;

console.log('JavaScript');
let time_start = new Date().getTime();
for (var num = 0; num <= limit; num++) 
    if (isPrime(num)) noPrimes++;
let time_end = new Date().getTime();
let delta_time = (time_end - time_start) / 1000;
console.log("Number of primes in the interval [" + min + "," + limit + "]: " + noPrimes);
console.log('Execution time: ' + delta_time + ' seconds');

console.log('Sieve of Eratosthenes');
time_start = new Date().getTime();
const primes = eratosthenes(limit);
time_end = new Date().getTime();
delta_time = (time_end - time_start) / 1000;
console.log("Number of primes in the interval [" + min + "," + limit + "]: " + primes.length);
console.log('Execution time: ' + delta_time + ' seconds');
//console.log(primes);

File: primes.rs
fn is_prime(n: u64) -> bool {
    if n <= 1 { return false; }
    if n <= 3 { return true;  }
    if (n % 2 == 0) || (n % 3 == 0) { return false; }
    let mut m = 5;
    while m <= (n as f64).sqrt() as u64 {
        if (n % m == 0) || (n % (m + 2) == 0) { return false; }
        m += 6;
    }
    true
}

fn main() {
    use std::time::Instant;

    println!("Rust");

	let mut count = 0;
    let limitmin = 0;
    let limitmax = 20000000;
    let now = Instant::now();
    for n in limitmin..limitmax + 1 {
        if is_prime(n) {
			count+=1;
        }
    }
    let elapsed = now.elapsed();
    let sec = (elapsed.as_secs() as f64) + (elapsed.subsec_nanos() as f64 / 1000_000_000.0);
    println!("Number of primes in the interval [{0},{1}]: {2}", limitmin, limitmax, count);
    println!("Time elapsed: {} seconds", sec);
}
